apps/passport-server/test/zuconnect.spec.ts

```diff
@@ -0,0 +1,242 @@
+
+  it("should sync with good API response", async () => {
+    await zuconnectTripshaSyncService.sync();
+    const tickets = await fetchAllZuconnectTickets(db);
+    expect(tickets.length).to.eq(5);
+  });
+
+  it("should soft-delete when tickets do not appear in API response", async () => {
+    server.use(makeHandler({ tickets: goodResponse.tickets.slice(0, 4) }));
+    await zuconnectTripshaSyncService.sync();
+    const tickets = await fetchAllZuconnectTickets(db);
+    expect(tickets.length).to.eq(4);
+
+    const deleted = await sqlQuery(
+      db,
+      `SELECT * FROM zuconnect_tickets WHERE is_deleted = TRUE`
+    );
+    expect(deleted.rowCount).to.eq(1);
+    expect(deleted.rows[0].external_ticket_id).to.eq(
+      goodResponse.tickets[4].id
+    );
+  });
+
+  it("soft-deleted ticket should be un-deleted if ticket appears again in API response", async () => {
+    await zuconnectTripshaSyncService.sync();
+    const tickets = await fetchAllZuconnectTickets(db);
+    expect(tickets.length).to.eq(5);
+
+    const deleted = await sqlQuery(
+      db,
+      `SELECT * FROM zuconnect_tickets WHERE is_deleted = TRUE`
+    );
+    expect(deleted.rowCount).to.eq(0);
+  });
+
+  it("should fail to sync with bad API response", async () => {
+    server.use(makeHandler(badEmptyResponse));
+    try {
+      expect(await zuconnectTripshaSyncService.sync()).to.throw;
+    } catch (e) {
+      /**
+       * We get detailed validation errors here due to the schema defined in
+       * {@link ZuconnectTripshaSchema}
+       */
+      expect((e as any).issues[0].code).to.eq("invalid_type");
+      expect((e as any).issues[0].path).to.deep.eq(["tickets"]);
+      expect((e as any).issues[0].expected).to.eq("array");
+      expect((e as any).issues[0].received).to.eq("undefined");
+    }
+
+    server.use(makeHandler(badTicketsResponse));
+    try {
+      expect(await zuconnectTripshaSyncService.sync()).to.throw;
+    } catch (e) {
+      expect((e as any).issues[0].code).to.eq("invalid_type");
+      expect((e as any).issues[0].path).to.deep.eq(["tickets", 0, "email"]);
+      expect((e as any).issues[0].expected).to.eq("string");
+      expect((e as any).issues[0].received).to.eq("undefined");
+    }
+  });
+
+  it("should be able to log in", async function () {
+    const result = await testLogin(application, goodResponse.tickets[0].email, {
+      expectEmailIncorrect: false,
+      expectUserAlreadyLoggedIn: false,
+      force: false,
+      skipSetupPassword: false
+    });
+
+    if (!result) {
+      throw new Error("failed to log in");
+    }
+
+    expect(result).to.not.be.undefined;
+
+    identity = result.identity;
+    user = result.user;
+  });
+
+  it("after a Zuconnect resident logs in, they should show up in the resident semaphore group and no other groups", async function () {
+    if (!user) {
+      throw new Error("expected user");
+    }
+
+    await application.services.semaphoreService.reload();
+
+    await sleep(100);
```

Comment:
idt you need this sleep here

Reply:
Good point, this was left over when copying from another test which waited for the semaphore service to do a scheduled reload, whereas here I'm doing it synchronously.